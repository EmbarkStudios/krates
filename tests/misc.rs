mod util;

#[test]
fn iter_names() {
    let contents = std::fs::read_to_string("tests/all-features.json")
        .map_err(|e| format!("failed to load metadata file: {}", e))
        .unwrap();

    let md: krates::cm::Metadata = serde_json::from_str(&contents)
        .map_err(|e| format!("failed to deserialize metadata: {}", e))
        .unwrap();

    let krates: krates::Krates = krates::Builder::new()
        .build_with_metadata(md, |_| {})
        .unwrap();

    let mut iter = krates.krates_by_name("winapi");

    let win28 = iter.next().unwrap();
    assert_eq!(win28.krate.name, "winapi");
    assert_eq!(
        win28.krate.version,
        krates::semver::Version::parse("0.2.8").unwrap()
    );

    let win38 = iter.next().unwrap();
    assert_eq!(win38.krate.name, "winapi");
    assert_eq!(
        win38.krate.version,
        krates::semver::Version::parse("0.3.9").unwrap()
    );

    assert!(iter.next().is_none());

    let mut iter = krates.krates_by_name("a");

    let a = iter.next().unwrap();
    assert_eq!(a.krate.name, "a");
    assert_eq!(
        a.krate.version,
        krates::semver::Version::parse("0.1.0").unwrap()
    );

    assert!(iter.next().is_none());

    let mut iter = krates.krates_by_name("winapi-x86_64-pc-windows-gnu");

    let wingnu = iter.next().unwrap();
    assert_eq!(wingnu.krate.name, "winapi-x86_64-pc-windows-gnu");
    assert_eq!(
        wingnu.krate.version,
        krates::semver::Version::parse("0.4.0").unwrap()
    );

    assert!(iter.next().is_none());
}

#[test]
fn iter_matches() {
    let contents = std::fs::read_to_string("tests/all-features.json")
        .map_err(|e| format!("failed to load metadata file: {}", e))
        .unwrap();

    let md: krates::cm::Metadata = serde_json::from_str(&contents)
        .map_err(|e| format!("failed to deserialize metadata: {}", e))
        .unwrap();

    let krates: krates::Krates = krates::Builder::new()
        .build_with_metadata(md, |_| {})
        .unwrap();

    {
        let any = krates::semver::VersionReq::STAR;
        let mut iter = krates.search_matches("winapi", any);

        let win28 = iter.next().unwrap();
        assert_eq!(win28.krate.name, "winapi");
        assert_eq!(
            win28.krate.version,
            krates::semver::Version::parse("0.2.8").unwrap()
        );

        let win38 = iter.next().unwrap();
        assert_eq!(win38.krate.name, "winapi");
        assert_eq!(
            win38.krate.version,
            krates::semver::Version::parse("0.3.9").unwrap()
        );

        assert!(iter.next().is_none());
    }

    {
        let two = krates::semver::VersionReq::parse("=0.2").unwrap();
        let mut iter = krates.search_matches("winapi", two);

        let win28 = iter.next().unwrap();
        assert_eq!(win28.krate.name, "winapi");
        assert_eq!(
            win28.krate.version,
            krates::semver::Version::parse("0.2.8").unwrap()
        );

        assert!(iter.next().is_none());
    }

    {
        let grtr = krates::semver::VersionReq::parse(">0.2.8").unwrap();
        let mut iter = krates.search_matches("winapi", grtr);

        let win38 = iter.next().unwrap();
        assert_eq!(win38.krate.name, "winapi");
        assert_eq!(
            win38.krate.version,
            krates::semver::Version::parse("0.3.9").unwrap()
        );

        assert!(iter.next().is_none());
    }

    {
        let none = krates::semver::VersionReq::parse("=0.4").unwrap();
        let mut iter = krates.search_matches("winapi", none);

        assert!(iter.next().is_none());
    }
}

#[test]
fn direct_dependents() {
    let mut kb = krates::Builder::new();
    kb.include_targets(std::iter::once((
        krates::cfg_expr::targets::get_builtin_target_by_triple("x86_64-unknown-linux-gnu")
            .unwrap()
            .triple
            .clone(),
        vec![],
    )));

    let grafs = util::build("direct.json", kb).unwrap();

    let id = grafs
        .actual
        .krates()
        .find(|k| k.0.repr.starts_with("reqwest"))
        .unwrap();

    let mut ids: Vec<_> = grafs
        .actual
        .direct_dependents(grafs.actual.nid_for_kid(&id.0).unwrap())
        .into_iter()
        .map(|jid| jid.krate.0.repr.as_str())
        .collect();

    ids.sort();
    let dd = ids.join("\n");

    insta::assert_snapshot!(dd);
}

#[test]
fn direct_dependencies() {
    let mut kb = krates::Builder::new();
    kb.include_targets(std::iter::once((
        krates::cfg_expr::targets::get_builtin_target_by_triple("x86_64-unknown-linux-gnu")
            .unwrap()
            .triple
            .clone(),
        vec![],
    )));

    let grafs = util::build("direct.json", kb).unwrap();

    let id = grafs
        .actual
        .krates()
        .find(|k| k.0.repr.starts_with("reqwest"))
        .unwrap();

    let mut ids: Vec<_> = grafs
        .actual
        .direct_dependencies(grafs.actual.nid_for_kid(&id.0).unwrap())
        .into_iter()
        .map(|jid| jid.krate.0.repr.as_str())
        .collect();

    ids.sort();
    let dd = ids.join("\n");

    insta::assert_snapshot!(dd);
}

#[test]
#[cfg(feature = "with-crates-index")]
fn bug_repro() {
    let mut kb = krates::Builder::new();
    kb.with_crates_io_index(None, krates::index::IndexKind::Sparse)
        .unwrap();

    let grafs = util::build("bug.json", kb).unwrap();

    insta::assert_snapshot!(grafs.dotgraph());
}
