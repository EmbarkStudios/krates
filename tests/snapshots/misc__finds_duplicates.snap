---
source: tests/misc.rs
expression: opaque
---
digraph {
    0 [ label = "crate bitflags 2.4.2" ]
    1 [ label = "crate bumpalo 3.14.0" ]
    2 [ label = "crate bytes 1.5.0" ]
    3 [ label = "crate cfg-if 1.0.0" ]
    4 [ label = "crate fnv 1.0.7" ]
    5 [ label = "crate futures-core 0.3.30" ]
    6 [ label = "crate futures-task 0.3.30" ]
    7 [ label = "crate futures-util 0.3.30" ]
    8 [ label = "crate getrandom 0.1.16" ]
    9 [ label = "crate getrandom 0.2.12" ]
    10 [ label = "crate http 0.2.11" ]
    11 [ label = "crate http 1.0.0" ]
    12 [ label = "crate http-body 0.4.6" ]
    13 [ label = "crate http-body 1.0.0" ]
    14 [ label = "crate http-body-util 0.1.0" ]
    15 [ label = "crate http-range-header 0.3.1" ]
    16 [ label = "crate itoa 1.0.10" ]
    17 [ label = "crate js-sys 0.3.67" ]
    18 [ label = "crate libc 0.2.152" ]
    19 [ label = "crate log 0.4.20" ]
    20 [ label = "crate once_cell 1.19.0" ]
    21 [ label = "crate pid 0.1.0 path+file:///krates/tests/pid" ]
    22 [ label = "crate pin-project-lite 0.2.13" ]
    23 [ label = "crate pin-utils 0.1.0" ]
    24 [ label = "crate proc-macro2 1.0.78" ]
    25 [ label = "crate quote 1.0.35" ]
    26 [ label = "crate syn 2.0.48" ]
    27 [ label = "crate tower-http 0.4.4" ]
    28 [ label = "crate tower-http 0.5.1" ]
    29 [ label = "crate tower-layer 0.3.2" ]
    30 [ label = "crate tower-service 0.3.2" ]
    31 [ label = "crate unicode-ident 1.0.12" ]
    32 [ label = "crate wasi 0.11.0+wasi-snapshot-preview1" ]
    33 [ label = "crate wasi 0.9.0+wasi-snapshot-preview1" ]
    34 [ label = "crate wasm-bindgen 0.2.90" ]
    35 [ label = "crate wasm-bindgen-backend 0.2.90" ]
    36 [ label = "crate wasm-bindgen-macro 0.2.90" ]
    37 [ label = "crate wasm-bindgen-macro-support 0.2.90" ]
    38 [ label = "crate wasm-bindgen-shared 0.2.90" ]
    39 [ label = "feature alloc" ]
    40 [ label = "feature alloc" ]
    41 [ label = "feature default" ]
    42 [ label = "feature default" ]
    43 [ label = "feature default" ]
    44 [ label = "feature default" ]
    45 [ label = "feature default" ]
    46 [ label = "feature alloc" ]
    47 [ label = "feature wasm-bindgen" ]
    48 [ label = "feature js" ]
    49 [ label = "feature default" ]
    50 [ label = "feature default" ]
    51 [ label = "feature sensitive-headers" ]
    52 [ label = "feature default" ]
    53 [ label = "feature proc-macro" ]
    54 [ label = "feature proc-macro" ]
    55 [ label = "feature default" ]
    56 [ label = "feature spans" ]
    57 [ label = "feature default" ]
    58 [ label = "feature default" ]
    59 [ label = "feature default" ]
    60 [ label = "feature default" ]
    61 [ label = "feature full" ]
    62 [ label = "feature default" ]
    63 [ label = "feature spans" ]
    64 [ label = "feature visit" ]
    65 [ label = "feature spans" ]
    66 [ label = "feature std" ]
    67 [ label = "feature std" ]
    68 [ label = "feature std" ]
    69 [ label = "feature bindgen" ]
    70 [ label = "feature js-sys" ]
    71 [ label = "feature bindgen" ]
    72 [ label = "feature wasm-bindgen" ]
    73 [ label = "feature js-sys" ]
    74 [ label = "feature std" ]
    75 [ label = "feature std" ]
    76 [ label = "feature std" ]
    77 [ label = "feature alloc" ]
    78 [ label = "feature race" ]
    79 [ label = "feature quote" ]
    80 [ label = "feature proc-macro" ]
    81 [ label = "feature printing" ]
    82 [ label = "feature parsing" ]
    83 [ label = "feature derive" ]
    84 [ label = "feature clone-impls" ]
    85 [ label = "feature std" ]
    86 [ label = "feature std" ]
    87 [ label = "feature spans" ]
    7 -> 5 [ label = "" ]
    7 -> 39 [ label = "" ]
    7 -> 6 [ label = "" ]
    7 -> 40 [ label = "" ]
    7 -> 22 [ label = "" ]
    7 -> 23 [ label = "" ]
    8 -> 3 [ label = "" ]
    8 -> 17 [ label = " 'wasm32-unknown-unknown'" ]
    8 -> 18 [ label = " 'cfg(unix)'" ]
    8 -> 41 [ label = " 'cfg(target_os = \"wasi\")'" ]
    8 -> 42 [ label = " 'wasm32-unknown-unknown'" ]
    9 -> 3 [ label = "" ]
    9 -> 17 [ label = " 'cfg(all(any(target_arch = \"wasm32\", target_arch = \"wasm64\"), target_os = \"unknown\"))'" ]
    9 -> 18 [ label = " 'cfg(unix)'" ]
    9 -> 32 [ label = " 'cfg(target_os = \"wasi\")'" ]
    9 -> 34 [ label = " 'cfg(all(any(target_arch = \"wasm32\", target_arch = \"wasm64\"), target_os = \"unknown\"))'" ]
    10 -> 43 [ label = "" ]
    10 -> 44 [ label = "" ]
    10 -> 16 [ label = "" ]
    11 -> 43 [ label = "" ]
    11 -> 44 [ label = "" ]
    11 -> 16 [ label = "" ]
    12 -> 43 [ label = "" ]
    12 -> 10 [ label = "" ]
    12 -> 22 [ label = "" ]
    13 -> 43 [ label = "" ]
    13 -> 45 [ label = "" ]
    14 -> 43 [ label = "" ]
    14 -> 46 [ label = "" ]
    14 -> 45 [ label = "" ]
    14 -> 13 [ label = "" ]
    14 -> 22 [ label = "" ]
    17 -> 42 [ label = "" ]
    21 -> 47 [ label = "" ]
    21 -> 48 [ label = "" ]
    21 -> 49 [ label = "" ]
    21 -> 50 [ label = "" ]
    21 -> 51 [ label = "" ]
    21 -> 52 [ label = "" ]
    24 -> 31 [ label = "" ]
    25 -> 24 [ label = "" ]
    25 -> 53 [ label = "" ]
    26 -> 24 [ label = "" ]
    26 -> 53 [ label = "" ]
    26 -> 25 [ label = "" ]
    26 -> 54 [ label = "" ]
    26 -> 31 [ label = "" ]
    27 -> 0 [ label = "" ]
    27 -> 43 [ label = "" ]
    27 -> 55 [ label = "" ]
    27 -> 7 [ label = "" ]
    27 -> 10 [ label = "" ]
    27 -> 12 [ label = "" ]
    27 -> 15 [ label = "" ]
    27 -> 22 [ label = "" ]
    27 -> 29 [ label = "" ]
    27 -> 30 [ label = "" ]
    28 -> 0 [ label = "" ]
    28 -> 43 [ label = "" ]
    28 -> 45 [ label = "" ]
    28 -> 13 [ label = "" ]
    28 -> 14 [ label = "" ]
    28 -> 22 [ label = "" ]
    28 -> 29 [ label = "" ]
    28 -> 30 [ label = "" ]
    34 -> 3 [ label = "" ]
    34 -> 36 [ label = "" ]
    34 -> 56 [ label = "" ]
    35 -> 57 [ label = "" ]
    35 -> 19 [ label = "" ]
    35 -> 58 [ label = "" ]
    35 -> 59 [ label = "" ]
    35 -> 60 [ label = "" ]
    35 -> 61 [ label = "" ]
    35 -> 62 [ label = "" ]
    35 -> 38 [ label = "" ]
    36 -> 60 [ label = "" ]
    36 -> 37 [ label = "" ]
    36 -> 63 [ label = "" ]
    37 -> 59 [ label = "" ]
    37 -> 60 [ label = "" ]
    37 -> 64 [ label = "" ]
    37 -> 61 [ label = "" ]
    37 -> 62 [ label = "" ]
    37 -> 35 [ label = "" ]
    37 -> 65 [ label = "" ]
    37 -> 38 [ label = "" ]
    57 -> 1 [ label = "" ]
    66 -> 2 [ label = "" ]
    43 -> 2 [ label = "" ]
    43 -> 66 [ label = "" ]
    67 -> 4 [ label = "" ]
    44 -> 4 [ label = "" ]
    44 -> 67 [ label = "" ]
    68 -> 5 [ label = "" ]
    68 -> 39 [ label = "" ]
    55 -> 5 [ label = "" ]
    55 -> 68 [ label = "" ]
    39 -> 5 [ label = "" ]
    40 -> 6 [ label = "" ]
    46 -> 7 [ label = "" ]
    46 -> 39 [ label = "" ]
    46 -> 40 [ label = "" ]
    47 -> 8 [ label = "" ]
    47 -> 69 [ label = "" ]
    47 -> 70 [ label = "" ]
    70 -> 8 [ label = "" ]
    70 -> 17 [ label = "" ]
    69 -> 8 [ label = "" ]
    69 -> 71 [ label = "" ]
    72 -> 9 [ label = "" ]
    72 -> 34 [ label = "" ]
    73 -> 9 [ label = "" ]
    73 -> 17 [ label = "" ]
    48 -> 9 [ label = "" ]
    48 -> 72 [ label = "" ]
    48 -> 73 [ label = "" ]
    74 -> 11 [ label = "" ]
    45 -> 11 [ label = "" ]
    45 -> 74 [ label = "" ]
    75 -> 18 [ label = "" ]
    49 -> 18 [ label = "" ]
    49 -> 75 [ label = "" ]
    76 -> 20 [ label = "" ]
    76 -> 77 [ label = "" ]
    78 -> 20 [ label = "" ]
    58 -> 20 [ label = "" ]
    58 -> 76 [ label = "" ]
    77 -> 20 [ label = "" ]
    77 -> 78 [ label = "" ]
    53 -> 24 [ label = "" ]
    59 -> 24 [ label = "" ]
    59 -> 53 [ label = "" ]
    54 -> 25 [ label = "" ]
    54 -> 53 [ label = "" ]
    60 -> 25 [ label = "" ]
    60 -> 54 [ label = "" ]
    64 -> 26 [ label = "" ]
    79 -> 26 [ label = "" ]
    79 -> 25 [ label = "" ]
    80 -> 26 [ label = "" ]
    80 -> 53 [ label = "" ]
    80 -> 54 [ label = "" ]
    81 -> 26 [ label = "" ]
    81 -> 79 [ label = "" ]
    82 -> 26 [ label = "" ]
    61 -> 26 [ label = "" ]
    83 -> 26 [ label = "" ]
    62 -> 26 [ label = "" ]
    62 -> 83 [ label = "" ]
    62 -> 82 [ label = "" ]
    62 -> 81 [ label = "" ]
    62 -> 84 [ label = "" ]
    62 -> 80 [ label = "" ]
    84 -> 26 [ label = "" ]
    50 -> 27 [ label = "" ]
    51 -> 28 [ label = "" ]
    52 -> 28 [ label = "" ]
    85 -> 33 [ label = "" ]
    41 -> 33 [ label = "" ]
    41 -> 85 [ label = "" ]
    86 -> 34 [ label = "" ]
    87 -> 34 [ label = "" ]
    87 -> 56 [ label = "" ]
    42 -> 34 [ label = "" ]
    42 -> 87 [ label = "" ]
    42 -> 86 [ label = "" ]
    65 -> 35 [ label = "" ]
    56 -> 36 [ label = "" ]
    56 -> 63 [ label = "" ]
    63 -> 37 [ label = "" ]
    63 -> 65 [ label = "" ]
}

